rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -115,17 +134,7 @@ class ReplayRSpace[C, P, A, R, K](val store: IStore[C, P, A, K], val branch: Bra
             }
           case Some(comms) =>
             val commOrDataCandidates: Either[COMM, Seq[DataCandidate[C, R]]] =
-              comms
-                .iterator()
-                .asScala
-                .collectFirst {
-                  case commRef: COMM =>
-                    runMatcher(Some(commRef)) match {
-                      case Some(x) => Right(x)
-                      case None    => Left(commRef)
-                    }
-                }
-                .get
+              getCommOrDataCandidates(comms.iterator().asScala.toList)
```

Comment:
```scala
comms.toStream.map(x => runMatcher(Some(x))).collectFirst { case Right(x) => x }.getOrElse(comms.head) 
```
should work.

Reply:
(with caveats about `Left` discussed on discord)

We should also try to do the same in `produce`.